doc: outline plain-v0 provisioner (#71)
operator-framework/rukpak: provisioner/plain-v0/README.md

# Plain-v0 Provisioner

## Summary

> The `plain-v0` provisioner is a core rukpak provisioner that knows how to interact with `plain+v0` format bundles.

Reviewer: It seems confusing to me that most things use `plain-v0`, but we're calling the format `plain+v0`. Should we make it `plain-v0` everywhere for consistency?

Author: From what I understood we were doing that intentionally to separate between the provisioner type and the bundle format. Since we have the existing `registry+v1` format, calling this format `plain+v0` seems consistent.

> The `plain-v0` provisioner is a core rukpak provisioner that knows how to interact with `plain+v0` format bundles.

Reviewer 2: It may help to have a link here to https://github.com/operator-framework/rukpak/blob/main/README.md#provisionerclass for what the `rukpak provisioner` is.

Author: I agree we could link back to the README, but we will remove the ProvisionerClass from the top-level README at the moment since we have not implemented that API yet

The `plain+v0` bundles, or plain bundles, are simply container images containing a set of static kubernetes YAML
manifests in a given directory. For example, below is an example Dockerfile that builds a plain `plain+v0` bundle from a
directory containing static manifests.

```dockerfile
FROM scratch
LABEL core.rukpak.io.bundle.mediatype.v1=plain+v0
LABEL core.rukpak.io.bundle.manifests.v1=/manifests
LABEL core.rukpak.io.bundle.package.v1=my-package

> LABEL core.rukpak.io.bundle.version.v1=0.0.1

Reviewer 3: nit: can we lose this labels right now as we're not actively using them anywhere in the provisioner?

> COPY ./manifests /manifests

Reviewer 3: can we add a note that we explicitly need the directory of manifests to live in the /manifests for the provisioner to work?

```

where the given `manifests` directory contains the kubernetes resources required to deploy an application, for example:

```
manifests
├── 00_namespace.yaml
├── 01_cluster_role.yaml
├── 02_role.yaml
├── 03_serviceaccount.yaml
├── 04_cluster_role_binding.yaml
├── 05_role_binding.yaml

> └── 06_deployment.yaml

Reviewer: This is _currently_ misleading since the apply ordering is not preserved. The plain-v0 provisioner currently delegates to helm's defined ordering.

Author: Yes, I removed the number prefix so as not to be misleading.

```

The `plain-v0` provisioner is able to unpack a given `plain+v0` bundle onto a cluster and then instantiate it, making
the content of the bundle available in the cluster. It does so by reconciling `Bundle` and `BundleInstance` types that
have the `spec.provisionerClassName` field set to `core.rukpak.io/plain-v0`. This field must be set to the correct
provisioner name in order for the `plain-v0` provisioner to see and interact with the bundle.

Below is an example of the provisioner reconciliation flow:

> ```mermaid

Reviewer 4: This is fantastic, really good to see here.

Reviewer 3: TIL that github has native support for this now. That's pretty neat.

graph TD
    C[Provisioner]
    C -->|Watches| D[Bundle]
    C -->|Watches| E[BundleInstance]
    D -->|References| F[Content]

>     E -->|Creates/Manages| F[Content]

Reviewer: At a high level, this flow makes sense, but the fact that there are arrows into the same content from both bundles and bundle instances could be confusing.

Technically, what's happening is that anytime the BI is reconciled it checks to see if `Content` is applied. If it isn't, it effectively creates a snapshot of the `Content` and applies that.

Author: Yes, I played around with some additional components to the chart, but fellback to something pretty simple at a high level. I think we could build this chart out to be more robust in a follow-up, I'd like to show the Bundle connected to backend storage and the BI connected to running components on-cluster

```

## Use cases

### Install and apply a specific version of a bundle

The `plain-v0` provisioner can install and make available a specific `plain+v0` bundle in the cluster.

Simply create a `Bundle` resource pointing to a specific version of your bundle, and a `BundleInstance` which references
that bundle. The `plain-v0` provisioner will unpack the provided Bundle onto the cluster, and eventually make the
content available on the cluster.

```yaml
apiVersion: core.rukpak.io/v1alpha1
kind: Bundle
metadata:
  name: my-bundle
spec:
  image: my-bundle@sha256:xyz123

>   provisionerClassName: core.rukpak.io/plain-v0

Reviewer 4: Something I've been confused by with this specific field - why is it "class name" wouldn't "provisioner" make more sense? This is more of a conversation fit for our slack channels but wanted to bring it up regardless.

Reviewer: The main reason is that there will likely be a `ProvisionerClass` API in the future. So this field will eventually be the name of another `ProvisionerClass` object, and _that_ object will contain the name of a specific provisioner along with the configuration that should be used with that provisioner.

That setup will allow the same provisioner to be used with different configurations by creating two separate `ProvisionerClass` objects.

Reviewer 4: Ah okay, I see that makes a lot more sense now. Thanks Joe - resolving.

---
apiVersion: core.rukpak.io/v1alpha1
kind: BundleInstance
metadata:
  name: my-bundle-instance
spec:
  provisionerClassName: core.rukpak.io/plain-v0
  bundleName: my-bundle
``` 

First, the Bundle will be in the Pending stage as the provisioner sees it and begins unpacking the referenced content:

```
NAME          IMAGE                                        PHASE     AGE
my-bundle     my-bundle@sha256:xyz123                      Pending   3s
```

Then eventually, as the bundle content is unpacked onto the cluster via the defined storage mechanism, the bundle status
will be updated to Unpacked, indicating that all its contents have been stored on-cluster.

```
NAME          IMAGE                                        PHASE      AGE

> my-bundle     my-bundle@sha256:xyz123                      Unpacked   13m

Reviewer 4: Does it normally take this long to unpack?

Reviewer: No. This is a good point. It might make sense to do `kubectl get bundles -w` and grab the relevant lines from that output to better reflect the time it takes.

In my tests, its usually just a few seconds (but its dependent on whether the image needs to be pulled and how big the image is)

Author: Yeah it doesn't, it was just sitting around on my cluster for a while 😄 good catch

```

Now that the bundle has been unpacked, the provisioner is able to create the resources in the bundle on the cluster.
These resources will be owned by the corresponding BundleInstance. Creating the BundleInstance on-cluster results in an
InstallationSucceeded Phase if the application of resources to the cluster was successful.

```
NAME                 DESIRED BUNDLE   INSTALLED BUNDLE   INSTALL STATE           AGE
my-bundle-instance   my-bundle        my-bundle          InstallationSucceeded   11s
```

### Make bundle content available but do not install it

There is a natural separation between sourcing of the content and application of that content via two separate RukPak
APIs, `Bundle` and `BundleInstance`. A user can specify a particular `Bundle` to be available in the cluster for
inspection before any application of the resources. Given a `Bundle` resource named `my-bundle`, the plain-v0

> provisioner will pull down and unpack the bundle to a variety of storage backends, such as ConfigMaps or
> PersistentVolumes.

Reviewer: ```suggestion
provisioner will pull down and unpack the bundle to a set of ConfigMaps.
```

> By default, `rukpak-system` is the default namespace for deploying plain-v0 provisioner-related system resources.

Reviewer 3: :+1:

> By default, `rukpak-system` is the default namespace for deploying plain-v0 provisioner-related system resources.

Reviewer 4: ```suggestion
By default, `rukpak-system` is the namespace for deploying plain-v0 provisioner-related system resources.
```

Surfacing the content of a bundle in a more user-friendly way, via a plugin or additional API, is on the RukPak roadmap.

### Pivoting between bundle versions

The `BundleInstance` API is meant to indicate the version of the bundle that should be active within the cluster. Given
two unpacked bundles in the cluster,
`my-bundle-v0.0.1` and `my-bundle-v0.0.2`, the `spec.bundleName` field of the related `BundleInstance` can be updated to
pivot to a different version. For example by pivoting from an older version, to the newer version, by updating
the `spec.bundleName` field the `plain-v0` provisioner will create the new bundle content on-cluster and remove the old
content. The provisioner also continually reconciles the created content via dynamic watches to ensure that all
resources referenced by the bundle are present on the cluster.

## Running locally

To experiment with the `plain-v0` provisioner locally, first setup a local cluster, or simply
have [kind](https://kind.sigs.k8s.io/) installed locally.

Once the cluster has been setup, take the following step from the root of the repository:

> * Run `make run` to build and deploy the provisioner onto the local cluster.

Reviewer 3: Ensure the repository has been cloned before running this?

From there, create some `Bundles` and `BundleInstance` types to see the provisioner in action. For an example bundle to
use, the [combo operator](https://github.com/operator-framework/combo) is a good candidate.

Create the bundle:

```yaml
apiVersion: core.rukpak.io/v1alpha1
kind: Bundle
metadata:
  name: combo-0.0.1
spec:
  image: quay.io/tyslaton/combo-plain-bundle:v0.0.1
  provisionerClassName: core.rukpak.io/plain-v0
```

Check the Bundle status via `kubectl get bundle combo-0.0.1`. Eventually the Bundle should show up as Unpacked.

```
NAME          IMAGE                                        PHASE      AGE
combo-0.0.1   quay.io/tyslaton/combo-plain-bundle:v0.0.1   Unpacked   17h
```

> Create the combo `BundleInstance` referencing the combo `Bundle` available in the cluster.

Reviewer 3: follow-up: mention we can create a set of Bundles/BundleInstances at the same time (e.g. it works with gitops and doesn't have any resource ordering requirements like OLM).

```yaml
apiVersion: core.rukpak.io/v1alpha1
kind: BundleInstance
metadata:
  name: combo-system

> spec:

Author: Good catch

  bundleName: combo-0.0.1
```

Check the BundleInstance status to ensure that the installation was successful.

```
NAME    DESIRED BUNDLE   INSTALLED BUNDLE   INSTALL STATE           AGE
combo   combo-0.0.1      combo-0.0.1        InstallationSucceeded   16h
```

From there, check out the combo operator deployment and ensure that the operator is present on the cluster.

```shell
kubectl -n combo get deployments.apps combo-operator -o yaml 
```

The operator should be successfully installed. To test out how the `plain-v0` provisioner continually reconciles
BundleInstance resources, try deleting the combo deployment.

```shell
kubectl -n combo delete deployments.apps combo-operator
```

Checking for the deployment again, it will be back on the cluster. The provisioner ensures that all resources required
for the BundleInstance to run are accounted for on-cluster.

To cleanup from the installation, simply remove the BundleInstance from the cluster. This will remove all references
resources including the deployment, RBAC, and the operator namespace. Then the Bundle can be removed as well, which
removes any unpacked manifests from the cluster.

> ```shell

Reviewer 4: 🎉
